cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -119,166 +124,420 @@ class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
  */
+trait QueueBasedTraverser extends Traverser {
+  /**
+   * Graph to be traversed.
+   */
+  def graph: Graph
 
-class BreadthFirstTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
-                            maxDepth: Option[Int], maxNumEdgesThresh: Option[Int], maxSteps: Long,
-                            onlyOnce: Boolean, prevNbrCounter: Option[PrevNbrCounter])
-    extends Traverser { self =>
+  /**
+   * Nodes to start the traversal from.
+   */
+  def homeNodeIds: Seq[Int]
 
-  private val log = Logger.get
-  // the number of items can be enqueued is bounded by maxSteps
-  // because we never need to enqueue more than maxSteps items
-  private var numEnqueuedEver = 0L
-  private val qu = new IntArrayFIFOQueue
+  /**
+   * Direction of the walk.
+   */
+  def dir: GraphDir
 
-  private val depthTracker = new IntInfoKeeper(true)
+  /**
+   * The priority of nodes adding to the queue. `BFS` adds nodes at the
+   * end and `DFS` adds nodes at the beginning.
+   */
+  object GraphTraverserNodePriority extends Enumeration {
+    type GraphTraverserNodePriority = Value
 
-  homeNodeIds foreach { id =>
-    depthTracker.recordInfo(id, 0)
-    enqueueNeighbors(id, 1)
+    val FIFO = Value
+    val LIFO = Value
   }
 
-  private def visitPotentialNode(id: Int, depth: Int) {
-    if (!(onlyOnce && seen(id))) {
-      qu.enqueue(id)
-      numEnqueuedEver += 1
+  import GraphTraverserNodePriority._
+
+  /**
+   * The priority of nodes adding to the queue.
+   */
+  def nodePriority: GraphTraverserNodePriority
+
+  /**
+   * Action performed when visiting node `node` (before iteratur retuns `node`).
+   * Should be implemented in subclass.
+   */
+  def visitNode(node: Node): Unit = {}
+
+  /**
+   * Queue that stores nodes to be visited next.
+   */
+  protected val queue = new IntArrayFIFOQueue()
+
+  /**
+   * Number of nodes ever enqueued in the `queue`.
+   */
+  protected var numEnqueuedEver = 0L
+
+  /**
+   * Optional counter of previous nodes to a given node.
+   */
+  lazy val prevNbrCounter: Option[PrevNbrCounter] = None
+
+  /**
+   * Depth of visit of a given node.
+   */
+  def depth(node: Int): Option[Int]
+
+  /**
+   * We mark nodes using 3 'colors': `Unenqueued`, `Enqueued`, `Visited`.
+   * Initially every node is implicitly marked with `Unenqueued`.
+   *
+   * Node is marked `Enqueued`, when we add it to the queue. We mark
+   * it `Visited` just before returning with iterator's `next` function.
+   */
+  lazy val coloring = new WalkNodeColoring {}
+
+  /**
+   * Checks if the node of a given color should be enqueued.
+   */
+  def shouldBeEnqueued(color: NodeColor.Color): Boolean
+
+  /**
+   * Options to limit traverser walk.
+   */
+  lazy val limits: WalkLimits = WalkLimits()
+
+  lazy val depthLimit: Limit = new Limit(limits.maxDepth)
+  lazy val degreeLimit: Limit = new Limit(limits.maxNumEdgesThreshold.map(_ + 1))
```

Comment:
The API before was so that degreeLimit was exclusive and depth inclusive. Changed the API. Now both are inclusive as limit word suggests (as for me at least)